### DISCUSSION TITLE
Partial data sync to RavenDB using AddOrPatch ?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#19095] [Q&A] [answered]
I'm doing a daily sync of employee data from an external system to RavenDB.

The goal is to sync(override) just some predefined properties of the Employee document, like name, address, working position etc.. and leave other properties untouched.

Data is pushed from the external system and temporarily gets bulk inserted into a collection.

Then the temporary collection is read and the `AddOrPatch` method is used to either insert a new document or update an existing one.

**The possible problem:**

- There are ~1500 employee records.
- I need to insert a new document or patch 10 properties if the document exists

As the `AddOrPatch` method takes just one property, I need to call that method for each property that needs to be patched, for each employee.

That results in a single query with `15000` patch instructions.

Can this be problematic in any way, and if yes is there a better solution to do that?

```
QUERY:
...
PATCH    empl/11540    this.GivenName = args.val_15021;    {"val_15021":"John"}
PATCH    empl/11540    this.FamilyName = args.val_15022;    {"val_15022":"Wick"}
PATCH    empl/11540    this.City = args.val_15023;    {"val_15023":"London"}
PATCH    empl/11540    this.PrimaryWorkPosition = args.val_15024;    {"val_15024":"workpos/ceo"}
PATCH    empl/2255    this.GivenName = args.val_15025;    {"val_15025":"Dr"}
PATCH    empl/2255    this.FamilyName = args.val_15026;    {"val_15026":"No"}
PATCH    empl/2255    this.City = args.val_15027;    {"val_15027":"New York"}
PATCH    empl/2255    this.PrimaryWorkPosition = args.val_15028;    {"val_15028":"workpos/ceo"}
...
```

**Comment by ayende** [comment, marked as answer]
If isn't _fun_ for us to process that on an individual basis, no.
It _works_, though. 

However, we have a better solution. Use `PatchCommandData` directly.

```
 s.Advanced.Defer(new PatchCommandData(empId, null, 
patch: new PatchRequest
 {
     Script = @"
// All properties set here
"
 }, patchIfMissing: newEmpInstanceHere));
```


So you can directly put all of them together.

**Reply by mcrio** [reply]
Thanks! It'll decrease the commands count to 1500 (1500 employees to sync). It still seems much to do at once, but it's much better.

I'll see at some point if I can introduce some batch processing and delays.

**Reply by mcrio** [reply]
hey @ayende  as the `patchIfMissing` is expecting another patch request, one way would be to write out all the properties in the script which is not  good in case in future more properties are added, or rely on reflection to populate that.

Is there some other more convenient way that would automatically add the document and the default metadata, or does this need to be manually specified as well?

something like this adds the document but the metadata would need to be added manually
```
          patchIfMissing: new PatchRequest
                    {
                        Script = @"Object.assign(this, args.Document);",
                        Values =
                        {
                            { "Document", newDocument },
                        },
                    }
```

**Reply by ayende** [reply]
You can see how we do it here:

https://github.com/ravendb/ravendb/blob/de22bdbdd23831e89d47dd7c2738f6dcd3845e0f/src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.Patch.cs#L160-L172

**Reply by mcrio** [reply]
Thanks. Nice, this works good, but it feels like it's buried to deep in code :) Next time I'll check the higher level method implementation for hints.

Just in case there ever will be time to work on the following improvements, as an average RavenDb client user:

- ...I'd like the documentation to say more about `CreateIfMissing` (nowhere mentioned) instead of `PatchIfMissing`. I'm not sure about the `PatchIfMIssinng` use-case, as there's nothing to patch, but I guess it's good to have for the scripting ability.
- ..I'd like the `PatchCommandData` to accept the document object and handle the conversion in the background, as we don't need to know what data needs to go into the metadata dictionary by default

Thanks again for the support.
